### PR TITLE
fix(substrait): remove optimize calls from substrait consumer

### PR DESCRIPTION
**Pull request by tokoko**
## Which issue does this PR close?
This is a necessary prerequisite for #12798

## Rationale for this change

substrait consumer should make the best effort for one-to-one traslation w/o invoking optimizer. Doing otherwise makes round-trip tests too complicated. 

## What changes are included in this PR?

When translating substrait ReadRel nodes, consumer constructs a dataframe first, applies projections with `select` and hopes that subsequent `into_optimized_plan` calls with push projections down to `TableScan`. In practice, this sometimes adds unnecessary projection nodes to the plan and also unnecessary "pushed down" projections to TableScan even when substrait doesn't specify any such thing.

This PR:
- removes `into_optimized_plan` calls from consumer.
- removes unnecessary projection nodes introduced in `ensure_schema_compatability`, instead opting to handcraft projections that are put into TableScan only when necessary (when base_schema provided in ReadRel has fewer fields than the actual table schema).
- Also fixes a bug during schema comparison in ensure_schema_compatability where comparison never yielded true because qualified and unqualified schemas were being compared. This led to consumer adding unnecessary projections to TableScan even when schemas match. Had to manually remove these projections from test assertions (incl. tpch queries).

## Are these changes tested?

This is essentially a refactor, being covered by the existing tests. I had to alter some asserts in schema compatibility tests as expected assertions were less than ideal in the first place (duplicated projections)

## Are there any user-facing changes?

minor, substrait plans are functionally the same, but may lack unnecessary projections.

### CODE REVIEW

**Comment by alamb** [comment]
cc @Blizzara  @westonpace and @vbarua

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:993]: Ah, the substrait_schema was produced with 
```
from_substrait_named_struct(named_struct, extensions)?
    .replace_qualifier(table_reference.clone());
 ```
so it includes the qualifiers, and that's why this wasn't matching. 

I have a small preference to **not add** the qualifiers to the substrait schema, and then **do** strip them here (given this is a simple read, the qualifier itself _should_ always match unless we've screwed something up, so given we need to either add it to one side or remove from another, I think it'd make more sense to do the comparison **without** it rather than with it. But doesn't matter much either way.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:998]: nit: move this up and then do Ok(t) inside the `if`?

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:786]: I believe with these changes the `extract_projection` may fail (or screw up) the plan if it is present on a case where ensure_schema_compatibility also adds a project into the table, since it'll override it: https://github.com/apache/datafusion/blob/60623afe6e2f44e1312c03b0eb1b06f622097f65/datafusion/substrait/src/logical_plan/consumer.rs#L344. (That's I think why the ensure_schema_compatibility originally had the extra projects and optimize call.) So we'll need to make the `extract_projection` a bit smarter so that it handles a project-of-a-project, I think.

We should also add a test for that case if possible, since I don't think we have any atm..

**Comment by Blizzara** [comment]
Thanks @tokoko - I think this makes sense overall and is the right thing to do, we'll just need to ensure the extract_projection doesn't break as a result :)

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs:993]: yeah, that makes more sense. let me give it a try

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs:786]: I don't think this is true, actually. Even before the changes, an optimize call would effectively do the same thing, it would push projections into TableScan that might have been overridden by extract_projection call afterwards.

it's actually safe to override them though, because projections provided in the substrait ReadRel are guaranteed to be the subset of the base_schema, right? In other words if ReadRel contains projections the number of clumns in base_schema no longer matters anyway.

P.S. Now that I think of it, we can actually get away with dumbing `extract_projection` down instead because now it will never be passed a `Projection` node, only ever a `TableScan`, so we no longer need to handle Projection case in the match.

Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/consumer.rs:786]: >  I don't think this is true, actually. Even before the changes, an optimize call would effectively do the same thing, it would push projections into TableScan that might have been overridden by extract_projection call afterwards.

Yeah, possible that it was buggy 😅 the idea I think was that the ensure_schema_compat would leave the Projection node there, and then extract_projection would trigger the second path rather than override the scan.project

> it's actually safe to override them though, because projections provided in the substrait ReadRel are guaranteed to be the subset of the base_schema, right?

Not quite, since the order / indexing might be different. I.e. if 
- the dataset has columns A, B, C, D (in that order), and 
- Substrait ReadRel's base_schema has C, B, A (in that order) + a project [0, 1]
then we should end up with a scan with a project of [3, 2], I think.

> P.S. Now that I think of it, we can actually get away with dumbing extract_projection down instead because now it will never be passed a Projection node, only ever a TableScan, so we no longer need to handle Projection case in the match.

Yup, that sounds right and a nice improvement!

Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs:786]: > Not quite, since the order / indexing might be different. I.e. if

yup, you're right. projected_schema will probably be fine, but projections themselves can be mixed up.

Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/consumer.rs:786]: > @Blizzara I'm thinking of switching gears a bit and curious what you think. I think the easiest solution here will be to first "apply" substrait projections to substrait base_schema and only then use `ensure_schema_compatability` with the resulting trimmed down schema as an input.

I think that'd be fine by me!

> One thing I'm unsure about is what to do when data types between table schema and a substrait base_schema do not match for a field which is projected away by substrait anyway. For example
> 
> * table has schema (a int, b str)
> * substrait base_schema is (a int, b int) # note that b has an incorrect type
> * substrait projections field is [0]
> 
> If I trim down base_schema first and only then check compatibility with the actual table, this won't throw an error, it will be as if column b was never mentioned in the plan at all. wdyt? Does that sound like the "right" behavior?

Also fine by me, given as long as the plan can safely execute (all columns used in any way exist and are of right type) then I don't see a need to fail that. But one could as well argue the whole included schema should be validated for consistency reasons, so dunno if @westonpace @vbarua have stricter opinions here?

> (On a related note, I think the same needs to applied to LocalFiles ReadType as well, which for some reason right now outright ignores base_schema. I think if we allow NamedTable base_schema to be different from the actual table schema, makes sense to do the same for LocalFiles).

Yeah, it wasn't done yet just to keep the PR smaller, but agreed the behavior should be same across all source types.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs:982]: I'm bumping into some round-trip errors when I try this. I suspect it should be related to optimization calls used in the tests themselves. I'll leave it as is for now and try to fix when I get rid of optimization calls in the tests as well.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs:984]: done

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs]: @Blizzara fixed with the approach outlined above. btw, turns out there was a [test](https://github.com/apache/datafusion/blob/c412c74a7e6874ee09b1749aec6fd2a0bc72faed/datafusion/substrait/tests/cases/substrait_validations.rs#L120) already that caught this, but apparently I did a classic rookie move in my initial commit and fixed the test instead of fixing the code :laughing:

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:855]: actually I guess all of this could be shared by all ReadRels? no need to do that refactor here if you don't want to but could be nice 🤷

File: datafusion/substrait/tests/cases/substrait_validations.rs
Comment (Blizzara) [inline datafusion/substrait/tests/cases/substrait_validations.rs:104]: nit, given you changed the order (to make the test more powerful, I think?):
```suggestion
            // the DataFusion schema { d, a, c, b } contains the Substrait schema { a, b, c }
```

**Review by Blizzara — COMMENTED** [review]
LGTM, I left couple more nits/ideas, feel free to look at them or disregard :)

File: datafusion/substrait/tests/cases/substrait_validations.rs
Comment (tokoko) [reply datafusion/substrait/tests/cases/substrait_validations.rs:104]: missed this. yes, before this, the naive wrong implementation yielded [b, a] which looked way too similar to [a, b]. now it will be [d, a].

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs:855]: yeah, this was deliberate to keep the PR smaller. I'll do the refactor separately if you don't mind. VirtualTable is an odd case, you don't really need schema compatibility check there, so better to leave that design to a different PR imho.

**Review by westonpace — APPROVED** [review]
Looks good to me, left a few comments (or perhaps more just random thoughts)

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs]: fixed, thanks

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs:1003]: yes, that's right. In addition there's a `ensure_field_compatability` check in there to check that individual data types match up.

...and also all this comes into play only if base_schema is different from catalog table schema. if they match (ln 986), we don't need a projection.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs:981]: the whole flow goes like this:
- `from_substrait_named_struct` gets substrait schema from the plan (base_schema)
- `apply_projection` applies MaskExpression from the plan to that schema. We're effectively combining base_schema and projection into a single schema object that we can use later to add projections to DF TableScan in one go.
- once we have that schema, `ensure_schema_compatability` does two things. 1) it checks that all requested columns exist and data types match up and 2) it adds projection to TableScan if not all columns from the table were requested.

How about I rename `apply_projection` to `apply_projection_to_schema` and `ensure_schema_compatability` to `adjust_with_requested_schema`?

@Blizzara any better ideas?

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs:349]: +1 to retaining metadata if it's possible, though I realize the existing code doesn't do this now either.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [inline datafusion/substrait/src/logical_plan/consumer.rs:328]: minor: if we're going to rename this, I might would `apply_masking` as that better corresponds to the fact that this is applying MaskExpressions specifically, and also makes it clearer that this isn't related to creating `Projects` / `Projections`.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [inline datafusion/substrait/src/logical_plan/consumer.rs:761]: Looking to future Substrait compatabilty. Substrait ReadRels have a [filter](https://github.com/substrait-io/substrait/blob/a0388ff69a83aa4addf51236b77bf275b3647590/proto/substrait/algebra.proto#L69) field that represents filter expressions that are applied _after_ the read but _before_ the masking (i.e. the filter expression can reference fields that are later masked out). 

Is the solution you're implementing here going to be compatible with that? By applying the masking here, we might potentially mask out fields that we're going to need.

Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs:761]: that's a good point. TableScan in DF seems to have a `filters` field, which I'm assuming is supposed to work the same way as substrait's filter, so I don't think we're in danger of doing that. Although this probably means that even the part of the base_schema that is being masked out needs some sort of validation because ReadRel's filter (and/or best_effort_filter) field might operate on them. This probably decides the question I raised above whether plans with inconsistent data types in masked out fields should be accepted or not.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs]: fixed

**Comment by tokoko** [comment]
I refactored a fair bit to address the comments.

- renamed `apply_projection` to `apply_masking`
- Split `ensure_schema_compatibility` into two functions. new `ensure_schema_compatibility` takes two schemas and validates them only w/o applying any sort of transformations. Also this is now happening before we call `apply_masking` to the substrait schema, the reason being that even if some of the fields are being masked out, ReadRel's `filter` and `best_effort_filter` expressions might still depend on them as @vbarua pointed out.
- The actual transform part from the original `ensure_schema_compatibility` function is now called `apply_projection` (i know... didn't have better ideas :laughing:) and is used to construct TableScan using masked substrait schema, no validations here anymore.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (vbarua) [inline datafusion/substrait/src/logical_plan/consumer.rs:1007]: I'm finding the code flow here a little tricky to follow with regards to how we would handle the ReadRel::filter.

1. This function is consuming a `substrait_schema` that has already had the mask expressions applied to it.
2. This function creates the DataFusion TableScan from the `table`, which is what we would need to attach the `ReadRel::filter` to using something like `scan.filters = <converted filters>`
3. But to build the `<converted filters>`, we'll need access to the unmasked `substrait_schema` in order to correctly resolve any Substrait column references. 

It might be easier to follow along with this, and make it easier to re-use, if we changed this function to something like:
```rust
/// Given the table produced by ReadRel::read_type, apply the remaining ReadRel fields
fn convert_read_rel(table: DataFrame, read_rel: &ReadRel) -> Result<LogicalPlan> {
    let named_struct = read.base_schema.as_ref().ok_or_else(|| {
        substrait_datafusion_err!("No base schema provided for ReadRel")
    })?;
    let substrait_schema =
        from_substrait_named_struct(named_struct, extensions)?
            .replace_qualifier(table_reference);

    ensure_schema_compatability(
        table.schema().to_owned(),
        substrait_schema.clone(),
    )?;

    let t = table.into_unoptimized_plan();

    // Apply filter (for later)
    // Apply best-effort filter (for later)
    // Apply masking (already done)
    // Apply RelCommon::emit (for later, and yes a ReadRel can have both masking AND a emit remapping)

   todo!()
}
```

Effectively the responsibility of the caller is to produce the underlying table, and then this function applies the rest of the ReadRel properties which are common to all of the different read_types.

All that being said, this might be more scope than you intended to take and might be better as a follow-up. I'll leave it up to you.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs:1043]: This only ever gets called with a df with a TableScan node so hopefully that default case should never be reached. I'll throw an error indicating it was called with something else.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs:1007]: > this might be more scope than you intended to take and might be better as a follow-up

I think I crossed that line a long ago, it's fine :laughing: 

Can you clarify how is this different from the current flow (aside from it being refactored into another function)? My thought was also to eventually add it [here](https://github.com/apache/datafusion/blob/e8b2b2a695e657034824200f9f2403c01179a3fd/datafusion/substrait/src/logical_plan/consumer.rs#L769) before calling `apply_masking` on the schema. We can always obtain a TableScan from a df and mutate it. then `apply_projection` would be changed to accept TableScan directly instead of a DataFrame.

Or maybe I'm misunderstanding your point. Are you challenging the fact that masking is applied to the schema first, which is then applied to a TableScan? The rationale there was that base_schema and masking are both needed in conjunction to come up with column indices and if I did in two different steps, I'd have to remap already added projections which seemed more error-prone.

Comment (vbarua) [reply datafusion/substrait/src/logical_plan/consumer.rs:1007]: Oh! You can attach a filter directly to the DataFrame? TIL. That addresses my data dependency concern, I thought you would only be able to apply the filter to the TableScan.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (tokoko) [reply datafusion/substrait/src/logical_plan/consumer.rs]: fixed

**Comment by tokoko** [comment]
@alamb can you force a rerun on the clippy job? seems like it failed on apt update. thanks

**Review by Blizzara — APPROVED** [review]
LGTM as well, thanks @tokoko for bearing with all the comments! :)

**Comment by alamb** [comment]
> @alamb just a reminder... this just needs clippy job to be rerun. Should I do a dummy commit?

Another trick I have found is to close the PR and then reopen it. I will do so

**Review by alamb — APPROVED** [review]
Thank you @tokoko  and @Blizzara @Blizzara and @westonpace for the reviews 🙏

**Comment by alamb** [comment]
> not sure how one gets access to it, but I think you should have a [re-run failed jobs](https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-workflow-runs/re-running-workflows-and-jobs) button just to re-run a single job w/o triggering the whole thing.

yes, you can do this if you are a committer. But the author can also close/reopen the PR

**Comment by alamb** [comment]
Thanks again @tokoko

**Pull request merged**
